AZP: Test release images with different CUDA versions

avildema opened the pull request:
## Why
Test UCX/Cuda binary compatibiloty

## How
Build UCX with Cuda 11.2, and test with Cuda 11.2 and 11.4

yosefe commented:
@avildema running dockers on GPU nodes causes other tests to fail because timeporary network devices are created and removed during the test
pls add `--net=nost` for all PRs which run dockers on GPU machines

avildema commented:
> @avildema running dockers on GPU nodes causes other tests to fail because timeporary network devices are created and removed during the test
> pls add `--net=nost` for all PRs which run dockers on GPU machines

added

avildema commented:
> @avildema running dockers on GPU nodes causes other tests to fail because timeporary network devices are created and removed during the test
> pls add `--net=nost` for all PRs which run dockers on GPU machines

impossible 

```
Error response from daemon: Container cannot be connected to network endpoints: nost, vsts_network_6071afc8fa88463d8291ddcce7b016b9
```

yosefe commented:
> > @avildema running dockers on GPU nodes causes other tests to fail because timeporary network devices are created and removed during the test
> > pls add `--net=nost` for all PRs which run dockers on GPU machines
> 
> impossible
> 
> ```
> Error response from daemon: Container cannot be connected to network endpoints: nost, vsts_network_6071afc8fa88463d8291ddcce7b016b9
> ```

so can you run on some other machine (not gpu01/02), or define that this test is "exclusive" (so no other test would run at the same time)?

avildema commented:
> > > @avildema running dockers on GPU nodes causes other tests to fail because timeporary network devices are created and removed during the test
> > > pls add `--net=nost` for all PRs which run dockers on GPU machines
> > 
> > 
> > impossible
> > ```
> > Error response from daemon: Container cannot be connected to network endpoints: nost, vsts_network_6071afc8fa88463d8291ddcce7b016b9
> > ```
> 
> so can you run on some other machine (not gpu01/02), or define that this test is "exclusive" (so no other test would run at the same time)?

other machine? I should build and run ucx+cuda , I can build on other , but it should be launch with gpu

yosefe commented:
> other machine? I should build and run ucx+cuda , I can build on other , but it should be launch with gpu

let's not run this PR until we find some solution since it causes other PRs to fail

avildema commented:
> > other machine? I should build and run ucx+cuda , I can build on other , but it should be launch with gpu
> 
> let's not run this PR until we find some solution since it causes other PRs to fail

may be need to create constant docker network and use it while I move gpu testing to docker?

yosefe commented:
> may be need to create constant docker network and use it while I move gpu testing to docker?

yes, this should solve the problem

avildema commented:
@yosefe can check

yosefe commented on buildlib/dockers/centos7-gpu.Dockerfile:
```diff
@@ -1,16 +1,15 @@
 ARG CUDA_VERSION
-FROM nvidia/cuda:${CUDA_VERSION}-devel-centos7
+ARG OS_VERSION
+FROM nvidia/cuda:${CUDA_VERSION}-devel-centos${OS_VERSION}
```
pls name this image "centos7-cuda" instead of "centos7-gpu"

yosefe commented on buildlib/dockers/docker-compose.yml:
```diff
@@ -12,6 +12,17 @@ services:
+  centos7-gpu-mofed5-cuda11:
+    image: centos7-gpu-mofed5-cuda11
+    build:
+      context: .
```
what does this do?

avildema replied:
for create empty docker image

yosefe commented on buildlib/dockers/docker-compose.yml:
```diff
@@ -12,6 +12,17 @@ services:
+    image: centos7-gpu-mofed5-cuda11
+    build:
+      context: .
+      network: host
```
pls comment that it's temporary solution to not disturb non-docker tests

avildema replied:
No, this is not applicable to CI, azure will run docker on its network 
it need more for manial launch

yosefe replied:
so you plane to remove it?

avildema replied:
No

yosefe commented on buildlib/dockers/docker-compose.yml:
```diff
@@ -12,6 +12,17 @@ services:
         MOFED_OS: rhel7.6
         CUDA_VERSION: 11.2.0
         OS_VERSION: 7
+  centos7-gpu-mofed5-cuda11:
```
centos7-cuda11-mofed5

avildema replied:
we already have this name

yosefe replied:
what is the difference in the container image then? maybe unite?

avildema replied:
added several addition links
```
RUN cd /usr/lib64 && \
    ln -s libudev.so.1 libudev.so && \
    ln -s libz.so.1 libz.so
```
and valgrind package

yosefe replied:
maybe it can be same container?
I think it's better to not have 2 unrelated containers for the same combination of os/cuda/mofed. it just creates confusion about which container should be used. can we add these links and valgrind package to the existing centos7-cuda11-mofed5?

avildema replied:
ok, no problems

yosefe commented on buildlib/pr/cuda_compatible.yml:
```diff
@@ -0,0 +1,114 @@
+          ./autogen.sh
+          mkdir pkg-build
+          cd pkg-build
+          ../contrib/configure-release --with-cuda --with-java=no --prefix=$(Build.Repository.LocalPath)/install
```
why need to disable java?

avildema replied:
can enable

yosefe commented on buildlib/pr/cuda_compatible.yml:
```diff
@@ -0,0 +1,114 @@
+      - bash: |
+          set -eEx
+          cd pkg-build
+          make -s -j 4
```
use `nproc` instead of 4

avildema replied:
`nproc` take all cores from the system

yosefe replied:
it takes according to agent process cpu affinity

yosefe commented on buildlib/pr/cuda_compatible.yml:
```diff
@@ -0,0 +1,114 @@
+          port=$((10000 + $(Build.BuildId)))
+          ./bin/ucx_perftest -t tag_bw -p $port &
+          server_pid=$!
+          sleep 1;
```
sleep 5

yosefe commented on buildlib/pr/main.yml:
```diff
@@ -38,9 +38,29 @@ resources:
       options: -v /hpc/local:/hpc/local -v /auto/sw_tools:/auto/sw_tools
+    - container: centos7_cuda11
+      image: rdmz-harbor.rdmz.labs.mlnx/ucx/centos7-mofed5-cuda11:1
+      options: -v /hpc/local:/hpc/local -v /auto/sw_tools:/auto/sw_tools --gpus all
```
should we also make rdma devices visible in docker?

yosefe commented:
@avildema as documented in https://github.com/openucx/ucx/wiki/Guidance-for-contributors, please do not force push during review
now, need to rebase on top of last reviewed commit - d0b2cbc8d4b25a0a419f11b8d003ad7c364eb6fa

avildema commented:
> @avildema as documented in https://github.com/openucx/ucx/wiki/Guidance-for-contributors, please do not force push during review
> now, need to rebase on top of last reviewed commit - [d0b2cbc](https://github.com/openucx/ucx/commit/d0b2cbc8d4b25a0a419f11b8d003ad7c364eb6fa)

done

yosefe commented on buildlib/pr/main.yml:
```diff
@@ -40,27 +40,22 @@ resources:
     - container: ubuntu20_cuda11
       image: rdmz-harbor.rdmz.labs.mlnx/ucx/ubuntu20.04-mofed5-cuda11:1
-      options: -v /hpc/local:/hpc/local -v /auto/sw_tools:/auto/sw_tools --gpus all
+      options: -v /hpc/local:/hpc/local -v /auto/sw_tools:/auto/sw_tools --gpus all --device=/dev/infiniband/uverbs1
```
1. why only uverbs1 and not also uverbs0?
2. add rdmacm
3. Is it possible to do something like `--device/dev/infiniband/*` ?

avildema replied:
1. added
2. added
3. docker: Error response from daemon: error gathering device information while adding custom device "/dev/infiniband/*": no such file or directory.

yosefe commented on buildlib/pr/cuda_compatible.yml:
```diff
@@ -0,0 +1,114 @@
+          ./bin/ucx_perftest -t tag_bw -p $port &
+          server_pid=$!
+          sleep 5
+          ./bin/ucx_perftest -t tag_bw -p $port 127.0.0.1 2>&1 | tee perf.txt
```
need to pass the following params to both client and server:
1. "-m cuda", to really test cuda memory
2. pls also add the following text file as "buildlib/pr/perftest_msg_pow2" and pass "-b buildlib/pr/perftest_msg_pow2" parameter. to test different message sizes

```
      8 -s       8 -n 20000
     16 -s      16 -n 20000
     32 -s      32 -n 20000
     64 -s      64 -n 20000
    128 -s     128 -n 14000
    256 -s     256 -n 7000
    512 -s     512 -n 3000
   1024 -s    1024 -n 2000
   2048 -s    2048 -n 1000
   4096 -s    4096 -n 1000
   8192 -s    8192 -n 800
  16384 -s   16384 -n 400
  32768 -s   32768 -n 200
  65536 -s   65536 -n 100
 131072 -s  131072 -n 50
 262144 -s  262144 -n 25
 524288 -s  524288 -n 12
1048576 -s 1048576 -n 10
2097152 -s 2097152 -n 10
```

yosefe commented:
@avildema pls squash without rebase

avildema commented:
GitHub close my PR. reopen https://github.com/openucx/ucx/pull/7453
